crates/matrix-sdk/src/config/client.rs

```diff
@@ -94,89 +97,44 @@ impl Debug for ClientConfig {
-        Ok(Box::new(StateStore::open().await?))
-    }
-
-    /// Open the indexeddb store at `name` (IndexedDB Database name), and open
-    /// the store with the given passphrase (if given) for encryption
-    pub async fn default_store_with_name(
-        name: &str,
-        passphrase: Option<&str>,
-    ) -> Result<Box<StateStore>> {
-        Ok(Box::new(match passphrase {
-            Some(pass) => StateStore::open_with_passphrase(name.to_owned(), pass).await?,
-            _ => StateStore::open_with_name(name.to_string()).await?,
-        }))
-    }
-}
-
-#[cfg(not(any(feature = "indexeddb_stores", feature = "sled_state_store")))]
-mod store_helpers {
-    use matrix_sdk_base::store::MemoryStore as StateStore;
-
-    use super::Result;
-    /// Open a new in-memory StateStore
-    pub async fn default_store() -> Result<Box<StateStore>> {
-        Ok(Box::new(StateStore::new()))
-    }
-
-    /// Alias for `default_store` - in Memory Stores are never named
-    pub async fn default_store_with_name(
-        _name: &str,
-        _passphrase: Option<&str>,
-    ) -> Result<Box<StateStore>> {
-        Ok(Box::new(StateStore::new()))
-    }
-}
-
-pub use store_helpers::{default_store, default_store_with_name};
-
 impl ClientConfig {
     /// Create a new default `ClientConfig`.
-    pub async fn new() -> Result<Self> {
-        let mut d = Self::default();
-        d.base_config = d.base_config.state_store(default_store().await?);
-        Ok(d)
+    pub fn new() -> Self {
+        Self::default()
     }
 
-    /// Create a new ClientConfig with a named state store, encrypted with the
-    /// given passphrase (if any)
-    pub async fn with_named_store(name: &str, passphrase: Option<&str>) -> Result<Self> {
-        let mut d = Self::default();
-        d.base_config = d.base_config.state_store(default_store_with_name(name, passphrase).await?);
-        Ok(d)
+    /// Create a new `ClientConfig` with the given [`StoreConfig`].
+    ///
+    /// The easiest way to get a [`StoreConfig`] is to use the [`make_config`]
+    /// method from the [`store`] module or directly from one of the store
+    /// crates.
+    ///
+    /// # Arguments
+    ///
+    /// * `store_config` - The configuration of the store.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # futures::executor::block_on(async {
+    /// # use matrix_sdk_base::store::MemoryStore;
+    /// # let custom_state_store = Box::new(MemoryStore::new());
+    /// use matrix_sdk::{Client, config::{ClientConfig, StoreConfig}};
+    ///
+    /// let store_config = StoreConfig::new().state_store(custom_state_store);
+    /// let client_config = ClientConfig::with_store_config(store_config)
+    ///     .use_discovery_response();
+    ///
+    /// # Result::<_, matrix_sdk::Error>::Ok(())
+    /// # });
+    /// ```
+    /// [`make_config`]: crate::store::make_config
+    /// [`store`]: crate::store
+    pub fn with_store_config(store_config: StoreConfig) -> Self {
```

Comment:
would it make more sense to have `store_config(mut self, store_config: StoreConfig)` instead of a second constructor? This way we would be consistent with other properties.

Reply:
I don't have strong feelings one way or the other but we already have the same behavior with other "builder" types.

Reply:
I'd prefer we'd aim for an actual builder type (something that finalizes with `.build()?` and consumes the builder, e.g. [derive_builder](https://docs.rs/derive_builder/latest/derive_builder/)), but for now, I prefer to have constructors like `Config::new_with_store(store);` rather than the more verbose `let mut config = Config::new(); config.set_something(something);`

My remark would be that it is a constructor and thus should start with `new_`....
```suggestion
    pub fn new_with_store_config(store_config: StoreConfig) -> Self {
```

Reply:
Having constructors named `with_x` isn't that uncommon. But in your second example, you seem to assume `&mut self` methods while we have `self` (consuming) methods on our builders. I would say that `Config::new().store_config(something)` would be the way to go. And if we want require users to *always* explicitly instantiate a store at some point, it will have to move out of `Config` entirely because that's optional itself.